// cronofy-elements/src/js/helpers/utils.CalendarSync.js
import React from "react";

export const parseAuthOptions = authorizationOptions => {
    return { ...authorizationOptions };
};

export const buildAuthUrl = ({
    authUrl,
    authOptions,
    provider,
    appDomain,
    singleProfile
}) => {
    let url;

    if (authUrl) {
        url = new URL(authUrl, "http://cronofy-default");
    } else {
        url = new URL(`${appDomain}/oauth/authorize`);

        // The "response_type" is always "code", so let's add that
        url.searchParams.append("response_type", "code");
    }

    // The "provider" is set internally, so let's add that
    url.searchParams.append("provider_name", provider);

    // If we're in singleProfile mode, set avoid_linking
    if (singleProfile) {
        url.searchParams.append("avoid_linking", true);
    }

    // Then add each of the authOptions as url params
    for (let param in authOptions) {
        if (authOptions.hasOwnProperty(param)) {
            url.searchParams.append(param, authOptions[param]);
        }
    }
    let link = url.href;

    if (link.startsWith("http://cronofy-default")) {
        link = link.replace("http://cronofy-default", "");
    }
    return link;
};

export const buildProviderData = (
    slug,
    name,
    auth,
    authUrl,
    appDomain,
    singleProfile,
    useBuildAuthUrl = buildAuthUrl
) => ({
    slug: slug,
    name: name,
    authlink: useBuildAuthUrl({
        authUrl,
        authOptions: auth,
        provider: slug,
        appDomain,
        singleProfile
    })
});

export const setupProviderData = (
    auth,
    appDomain,
    authUrl = false,
    singleProfile = false,
    useBuildProviderData = buildProviderData
) => [
    useBuildProviderData(
        "google",
        "Google",
        auth,
        authUrl,
        appDomain,
        singleProfile
    ),
    useBuildProviderData(
        "apple",
        "iCloud",
        auth,
        authUrl,
        appDomain,
        singleProfile
    ),
    useBuildProviderData(
        "office365",
        "Office 365",
        auth,
        authUrl,
        appDomain,
        singleProfile
    ),
    useBuildProviderData(
        "exchange",
        "Exchange",
        auth,
        authUrl,
        appDomain,
        singleProfile
    ),
    useBuildProviderData(
        "live_connect",
        "Outlook.com",
        auth,
        authUrl,
        appDomain,
        singleProfile
    )
];

export const getProviderIcon = (provider, fill) => {
    if (provider === "google") {
        if (fill) {
            return (
                <React.Fragment>
                    <defs>
                        <path d="M19.674 8.682h-9.318v3.863h5.364c-.5 2.455-2.591 3.864-5.364 3.864A5.897 5.897 0 0 1 4.447 10.5a5.897 5.897 0 0 1 5.91-5.91c1.408 0 2.68.5 3.68 1.32L16.948 3C15.174 1.455 12.902.5 10.357.5c-5.546 0-10 4.455-10 10s4.454 10 10 10c5 0 9.545-3.636 9.545-10 0-.59-.091-1.227-.228-1.818z" />
                        <path
                            d="M19.318 8.182H10v3.863h5.364c-.5 2.455-2.591 3.864-5.364 3.864A5.897 5.897 0 0 1 4.09 10 5.897 5.897 0 0 1 10 4.09c1.41 0 2.682.5 3.682 1.32L16.59 2.5C14.818.955 12.545 0 10 0 4.455 0 0 4.455 0 10s4.455 10 10 10c5 0 9.545-3.636 9.545-10 0-.59-.09-1.227-.227-1.818z"
                            id="google_a"
                        />
                        <path
                            d="M19.318 8.182H10v3.863h5.364c-.5 2.455-2.591 3.864-5.364 3.864A5.897 5.897 0 0 1 4.09 10 5.897 5.897 0 0 1 10 4.09c1.41 0 2.682.5 3.682 1.32L16.59 2.5C14.818.955 12.545 0 10 0 4.455 0 0 4.455 0 10s4.455 10 10 10c5 0 9.545-3.636 9.545-10 0-.59-.09-1.227-.227-1.818z"
                            id="google_c"
                        />
                        <path
                            d="M19.318 8.182H10v3.863h5.364c-.5 2.455-2.591 3.864-5.364 3.864A5.897 5.897 0 0 1 4.09 10 5.897 5.897 0 0 1 10 4.09c1.41 0 2.682.5 3.682 1.32L16.59 2.5C14.818.955 12.545 0 10 0 4.455 0 0 4.455 0 10s4.455 10 10 10c5 0 9.545-3.636 9.545-10 0-.59-.09-1.227-.227-1.818z"
                            id="google_e"
                        />
                        <path
                            d="M19.318 8.182H10v3.863h5.364c-.5 2.455-2.591 3.864-5.364 3.864A5.897 5.897 0 0 1 4.09 10 5.897 5.897 0 0 1 10 4.09c1.41 0 2.682.5 3.682 1.32L16.59 2.5C14.818.955 12.545 0 10 0 4.455 0 0 4.455 0 10s4.455 10 10 10c5 0 9.545-3.636 9.545-10 0-.59-.09-1.227-.227-1.818z"
                            id="google_g"
                        />
                    </defs>
                    <g fill="none" fillRule="evenodd">
                        <g transform="translate(0 .498)">
                            <mask id="google_b" fill="#fff">
                                <use xlinkHref="#google_a" />
                            </mask>
                            <path
                                fill="#FBBC05"
                                mask="url(#google_b)"
                                d="M-.91 15.91V4.09L6.819 10z"
                            />
                        </g>
                        <g transform="translate(0 .498)">
                            <mask id="google_d" fill="#fff">
                                <use xlinkHref="#google_c" />
                            </mask>
                            <path
                                fill="#EA4335"
                                fillRule="nonzero"
                                mask="url(#google_d)"
                                d="M-.91 4.09L6.819 10 10 7.227l10.91-1.772V-.91H-.91z"
                            />
                        </g>
                        <g transform="translate(0 .498)">
                            <mask id="google_f" fill="#fff">
                                <use xlinkHref="#google_e" />
                            </mask>
                            <path
                                fill="#34A853"
                                fillRule="nonzero"
                                mask="url(#google_f)"
                                d="M-.91 15.91L12.728 5.454l3.591.454L20.91-.909v21.818H-.909z"
                            />
                        </g>
                        <g transform="translate(0 .498)">
                            <mask id="google_h" fill="#fff">
                                <use xlinkHref="#google_g" />
                            </mask>
                            <path
                                fill="#4285F4"
                                fillRule="nonzero"
                                mask="url(#google_h)"
                                d="M20.91 20.91L6.817 10 5 8.636l15.91-4.545z"
                            />
                        </g>
                    </g>
                </React.Fragment>
            );
        } else {
            return (
                <path d="M19.674 8.682h-9.318v3.863h5.364c-.5 2.455-2.591 3.864-5.364 3.864A5.897 5.897 0 0 1 4.447 10.5a5.897 5.897 0 0 1 5.91-5.91c1.408 0 2.68.5 3.68 1.32L16.948 3C15.174 1.455 12.902.5 10.357.5c-5.546 0-10 4.455-10 10s4.454 10 10 10c5 0 9.545-3.636 9.545-10 0-.59-.091-1.227-.228-1.818z" />
            );
        }
    }
    if (provider === "live_connect") {
        if (fill) {
            return (
                <path
                    fill="#0071C5"
                    d="M11.591.54h1.385v4.633c2.4.004 4.797-.007 7.195 0 .472-.024 1.119.113 1.11.708.087 2.93.002 5.867.041 8.8.01.519-.007 1.336-.713 1.343-2.541.088-5.09.007-7.633.035v4.4H11.54c-3.53-.64-7.069-1.231-10.603-1.852C.934 13.203.934 7.799.934 2.397 4.487 1.776 8.041 1.173 11.591.54zm1.385 5.327v2.56c.906.845 1.8 1.705 2.72 2.536 1.528-1.434 3.053-2.87 4.586-4.299.255-.19.347-.488.391-.792-2.566-.011-5.13-.002-7.697-.005zm3.447 5.302c-.257.213-.512.54-.89.47-.917-.697-1.69-1.568-2.554-2.332-.005 2.018-.003 4.037-.003 6.057h7.644c-.002-2.715.005-5.431-.002-8.146-1.4 1.315-2.803 2.629-4.195 3.951zM5.9 6.715c-2.223.795-2.626 3.778-1.797 5.717.63 1.681 2.948 2.355 4.331 1.172 1.501-1.314 1.608-3.69.853-5.427-.554-1.263-2.08-1.955-3.387-1.462zm.225 1.392c.616-.294 1.387.005 1.677.614.463.917.46 2.036.157 3.004-.215.688-.968 1.256-1.697.985-.91-.288-1.186-1.367-1.196-2.212-.011-.876.19-1.967 1.059-2.39z"
                />
            );
        } else {
            return (
                <path d="M11.591.54h1.385v4.633c2.4.004 4.797-.007 7.195 0 .472-.024 1.119.113 1.11.708.087 2.93.002 5.867.041 8.8.01.519-.007 1.336-.713 1.343-2.541.088-5.09.007-7.633.035v4.4H11.54c-3.53-.64-7.069-1.231-10.603-1.852C.934 13.203.934 7.799.934 2.397 4.487 1.776 8.041 1.173 11.591.54zm1.385 5.327v2.56c.906.845 1.8 1.705 2.72 2.536 1.528-1.434 3.053-2.87 4.586-4.299.255-.19.347-.488.391-.792-2.566-.011-5.13-.002-7.697-.005zm3.447 5.302c-.257.213-.512.54-.89.47-.917-.697-1.69-1.568-2.554-2.332-.005 2.018-.003 4.037-.003 6.057h7.644c-.002-2.715.005-5.431-.002-8.146-1.4 1.315-2.803 2.629-4.195 3.951zM5.9 6.715c-2.223.795-2.626 3.778-1.797 5.717.63 1.681 2.948 2.355 4.331 1.172 1.501-1.314 1.608-3.69.853-5.427-.554-1.263-2.08-1.955-3.387-1.462zm.225 1.392c.616-.294 1.387.005 1.677.614.463.917.46 2.036.157 3.004-.215.688-.968 1.256-1.697.985-.91-.288-1.186-1.367-1.196-2.212-.011-.876.19-1.967 1.059-2.39z" />
            );
        }
    }
    if (provider === "apple") {
        return (
            <path d="M17.1 16.018c-.298.69-.652 1.326-1.061 1.91-.559.796-1.016 1.347-1.369 1.654-.546.502-1.131.76-1.758.774-.45 0-.993-.128-1.625-.388-.633-.258-1.216-.386-1.748-.386-.559 0-1.158.128-1.799.386-.641.26-1.158.395-1.554.409-.6.025-1.2-.24-1.798-.795-.382-.333-.86-.904-1.431-1.713-.614-.863-1.118-1.864-1.513-3.006-.424-1.233-.636-2.427-.636-3.582 0-1.324.286-2.466.86-3.423a5.04 5.04 0 0 1 1.799-1.82 4.84 4.84 0 0 1 2.432-.687c.478 0 1.104.148 1.882.438.776.291 1.274.439 1.492.439.164 0 .717-.173 1.655-.517.887-.32 1.636-.452 2.25-.4 1.662.135 2.91.79 3.741 1.97-1.487.901-2.222 2.163-2.207 3.781.013 1.26.47 2.31 1.37 3.143a4.5 4.5 0 0 0 1.368.897c-.11.319-.226.623-.35.916zM13.29 1.017c0 .988-.361 1.91-1.08 2.764-.87 1.016-1.92 1.602-3.058 1.51a3.076 3.076 0 0 1-.023-.375c0-.948.413-1.963 1.146-2.793.366-.42.832-.77 1.396-1.049.564-.274 1.097-.426 1.598-.452.014.132.02.264.02.395z" />
        );
    }
    if (provider === "office365") {
        if (fill) {
            return (
                <path
                    fill="#DC3E15"
                    d="M11.51 2.976c.031.192.039 14.085.01 14.388l-7.085-1.688V5.883L.788 4.458c-.007.005-.014.007-.02.011-.006.004-.013.01-.013.016-.004.036-.009.073-.009.11v11.863c.054.067.124.075.185.098l10.41 3.896c.12.045.229.067.362.028 1.904-.553 3.81-1.099 5.714-1.647l.095-.031V2.122l-1.632-.454c-1.381-.38-2.764-.758-4.144-1.142a.59.59 0 0 0-.386.014C7.937 1.8 4.523 3.058 1.11 4.316a.463.463 0 0 0-.19.113l10.59-1.453z"
                />
            );
        } else {
            return (
                <path d="M11.51 2.976c.031.192.039 14.085.01 14.388l-7.085-1.688V5.883L.788 4.458c-.007.005-.014.007-.02.011-.006.004-.013.01-.013.016-.004.036-.009.073-.009.11v11.863c.054.067.124.075.185.098l10.41 3.896c.12.045.229.067.362.028 1.904-.553 3.81-1.099 5.714-1.647l.095-.031V2.122l-1.632-.454c-1.381-.38-2.764-.758-4.144-1.142a.59.59 0 0 0-.386.014C7.937 1.8 4.523 3.058 1.11 4.316a.463.463 0 0 0-.19.113l10.59-1.453z" />
            );
        }
    }
    if (provider === "exchange") {
        if (fill) {
            return (
                <path
                    fill="#0071C5"
                    d="M11.772.701L.234 3.01v15.384l11.538 2.308v-20zm.77 4.616V6.71h.312c.078 0 .144-.006.144.072l1.01.985c.078.079.078.163 0 .24l-.385.385c-.078.079-.162.079-.24 0l-.457-.456h-.385v2.692c.691.616 1.313 1.067 1.082 1.298-.078.153-.544-.078-1.082-.385v2.765h.144c1.232-.923 2.242-2.614 2.549-2.308.384.307-2.074 3.558-2.38 3.558h-.313v1.298h6.683c.538 0 1.01-.472 1.01-1.01V6.326c0-.538-.472-1.01-1.01-1.01H12.54zm-4.688.312v1.683l-2.62.168v1.923l2.38-.096v1.707h-2.38v1.995l2.764.144v1.707l-4.375-.385v-8.39l4.23-.456zm8.149 1.082h2.62c.153 0 .216.063.216.216v2.476c0 .078.006.144-.072.144l-.985 1.01c-.079.078-.163.078-.24 0l-.385-.313c-.079-.078-.079-.162 0-.24l.456-.457v-1.61h-1.37c-1.232.922-2.241 2.614-2.548 2.307-.385-.306 2.001-3.455 2.308-3.533zm-.53 3.557c.61-.045 3.462 2.14 3.462 2.356v1.61h-.096v1.01c0 .154-.063.217-.216.217h-2.476c-.078 0-.144.006-.144-.072l-1.01-.986c-.078-.078-.078-.162 0-.24l.385-.385c.078-.078.162-.078.24 0l.457.457h1.61v-1.37c-.922-1.232-2.614-2.242-2.307-2.548a.141.141 0 0 1 .096-.049z"
                />
            );
        } else {
            return (
                <path d="M11.772.701L.234 3.01v15.384l11.538 2.308v-20zm.77 4.616V6.71h.312c.078 0 .144-.006.144.072l1.01.985c.078.079.078.163 0 .24l-.385.385c-.078.079-.162.079-.24 0l-.457-.456h-.385v2.692c.691.616 1.313 1.067 1.082 1.298-.078.153-.544-.078-1.082-.385v2.765h.144c1.232-.923 2.242-2.614 2.549-2.308.384.307-2.074 3.558-2.38 3.558h-.313v1.298h6.683c.538 0 1.01-.472 1.01-1.01V6.326c0-.538-.472-1.01-1.01-1.01H12.54zm-4.688.312v1.683l-2.62.168v1.923l2.38-.096v1.707h-2.38v1.995l2.764.144v1.707l-4.375-.385v-8.39l4.23-.456zm8.149 1.082h2.62c.153 0 .216.063.216.216v2.476c0 .078.006.144-.072.144l-.985 1.01c-.079.078-.163.078-.24 0l-.385-.313c-.079-.078-.079-.162 0-.24l.456-.457v-1.61h-1.37c-1.232.922-2.241 2.614-2.548 2.307-.385-.306 2.001-3.455 2.308-3.533zm-.53 3.557c.61-.045 3.462 2.14 3.462 2.356v1.61h-.096v1.01c0 .154-.063.217-.216.217h-2.476c-.078 0-.144.006-.144-.072l-1.01-.986c-.078-.078-.078-.162 0-.24l.385-.385c.078-.078.162-.078.24 0l.457.457h1.61v-1.37c-.922-1.232-2.614-2.242-2.307-2.548a.141.141 0 0 1 .096-.049z" />
            );
        }
    }
};
