// cronofy-elements/src/js/components/AvailabilityViewer/HoverSlot.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { StatusContext, ThemeContext } from "./AvailabilityViewer";

const HoverSlot = ({ available }) => {
    const [status, setStatus] = useContext(StatusContext);
    const [theme, setTheme] = useContext(ThemeContext);

    const classModifier =
        !available && status.slotSelection === "unrestricted"
            ? `${theme.prefix}__hover-slot--unavailable`
            : `${theme.prefix}__hover-slot--${
                  available ? "available" : "unavailable"
              }`;
    return (
        <div
            className={`${theme.prefix}__hover-slot ${classModifier}`}
            css={css`
                width: 100%;
                height: 100%;
                flex-grow: 1;
                position: relative;
                cursor: pointer;
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: center;
                background: ${available ||
                status.slotSelection === "unrestricted"
                    ? theme.colors.availableHover
                    : theme.colors.unavailableHover};
            `}
        />
    );
};

export default HoverSlot;
