// cronofy-elements/src/js/components/AvailabilityViewer/Loading.js
import React, { useContext } from "react";
import { css, jsx, keyframes } from "@emotion/core";
import { transparentize } from "polished";

import { I18nContext, ThemeContext } from "./AvailabilityViewer";

const Loading = () => {
    const [theme, setTheme] = useContext(ThemeContext);
    const i18n = useContext(I18nContext);

    const spin = keyframes`
        0% {
            transform: rotate(0deg);
        }
        100%{
            transform: rotate(360deg);
        }
    `;

    return (
        <div
            css={css`
                position: absolute;
                z-index: 5;
                top: 0;
                left: 0;
                width: 100%;
                height: 100%;
                background: ${transparentize(0.8, theme.colors.black)};
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: center;
            `}
            className={`${theme.prefix}__loading`}
        >
            <svg
                css={css`
                    width: 2em;
                    height: 2em;
                    fill: ${theme.colors.white};
                    animation: ${spin} infinite linear 1s;
                `}
                className={`${theme.prefix}__loading-icon`}
            >
                <use xlinkHref={`#reload`} />
            </svg>
            <span className={`${theme.prefix}__loading-text visuallyhidden`}>
                {i18n.t("loading")}
            </span>
        </div>
    );
};

export default Loading;
