// cronofy-elements/src/js/components/AvailabilityViewer/Preloading.js
import React, { useContext } from "react";
import { css, jsx, keyframes } from "@emotion/core";
import { transparentize } from "polished";

import { I18nContext, ThemeContext } from "./AvailabilityViewer";

const Preloading = () => {
    const [theme, setTheme] = useContext(ThemeContext);
    const i18n = useContext(I18nContext);

    const spin = keyframes`
        0% {
            transform: rotate(0deg);
        }
        100%{
            transform: rotate(360deg);
        }
    `;

    return (
        <div
            css={css`
                position: absolute;
                z-index: 5;
                top: 0px;
                left: 0px;
                width: 40px;
                height: 40px;
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: center;
            `}
            className={`${theme.prefix}__preloading`}
        >
            <svg
                css={css`
                    width: 20px;
                    height: 20px;
                    fill: ${theme.colors.black};
                    animation: ${spin} infinite linear 1s;
                `}
                className={`${theme.prefix}__preloading-icon`}
            >
                <use xlinkHref={`#reload`} />
            </svg>
            <span
                css={css`
                    position: absolute;
                    top: 50%;
                    left: 100%;
                    transform: translateY(-50%);
                    color: ${theme.colors.black};
                    line-height: 1;
                    font-size: 12px;
                `}
                className={`${theme.prefix}__preloading-text`}
            >
                {i18n.t("refreshing")}
            </span>
        </div>
    );
};

export default Preloading;
