// cronofy-elements/src/js/components/AvailabilityViewer/SelectionMask.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";
import { transparentize } from "polished";

import { ThemeContext } from "./AvailabilityViewer";
import { HoverContext, SelectionContext } from "./Week";

const SelectionMask = () => {
    const [selectedSlots, setSelectedSlots] = useContext(SelectionContext);
    const [hover, setHover] = useContext(HoverContext);
    const [theme, setTheme] = useContext(ThemeContext);

    const handleDeselection = () => {
        // If we setState on MouseMove for slots, the fans on a 2019 🍏 MBP will start to
        // spin (sounding like a helicopter). But *sometimes* we need this "realtime" mouse
        // position (i.e. when clearing a selection and reseting hover).
        setHover({ ...hover, helicopterMode: true });
        setSelectedSlots({});
    };

    return (
        <div
            css={css`
                position: absolute;
                top: 0;
                left: 0;
                width: 100%;
                height: 100%;
                background: ${transparentize(0.8, theme.colors.black)};
                z-index: 4;
                cursor: pointer;
            `}
            className={`${theme.prefix}__selection-mask`}
            onClick={handleDeselection}
        />
    );
};

export default SelectionMask;
