// cronofy-elements/src/js/components/AvailabilityViewer/TimeSelectorTrigger.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { ThemeContext } from "./AvailabilityViewer";

const TimeSelectorTrigger = ({ handleCalendarSelection }) => {
    const [theme, setTheme] = useContext(ThemeContext);
    return (
        <div
            css={css`
                position: absolute;
                bottom: 100%;
                right: 0;
                width: 20px;
                height: 20px;
                transform: translate(-10px, -10px);
                cursor: pointer;
                @media (max-width: 650px) {
                    display: none;
                }
            `}
            onClick={handleCalendarSelection}
            className={`${theme.prefix}__time-select__trigger`}
        >
            <svg
                width="20"
                height="20"
                xmlns="http://www.w3.org/2000/svg"
                className={`${theme.prefix}__time-select__trigger-icon`}
            >
                <path
                    d="M3.94 6.5L2.22 3.64l1.42-1.42L6.5 3.94c.52-.3 1.1-.54 1.7-.7L9 0h2l.8 3.24c.6.16 1.18.4 1.7.7l2.86-1.72 1.42 1.42-1.72 2.86c.3.52.54 1.1.7 1.7L20 9v2l-3.24.8c-.16.6-.4 1.18-.7 1.7l1.72 2.86-1.42 1.42-2.86-1.72c-.52.3-1.1.54-1.7.7L11 20H9l-.8-3.24c-.6-.16-1.18-.4-1.7-.7l-2.86 1.72-1.42-1.42 1.72-2.86c-.3-.52-.54-1.1-.7-1.7L0 11V9l3.24-.8c.16-.6.4-1.18.7-1.7zM10 13a3 3 0 1 0 0-6 3 3 0 0 0 0 6z"
                    fill={theme.colors.hairline}
                />
            </svg>
        </div>
    );
};

export default TimeSelectorTrigger;
