// cronofy-elements/src/js/components/AvailabilityViewer/Tooltip.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { ThemeContext } from "./AvailabilityViewer";

const Tooltip = ({ time, top = false, hover = true }) => {
    const [theme, setTheme] = useContext(ThemeContext);
    return (
        <span
            css={css`
                display: block;
                opacity: ${hover ? 1 : 0};
                position: absolute;
                width: 110%;
                min-width: 130px;
                height: 36px;
                color: ${theme.colors.white};
                display: block;
                ${top ? `top: 100%;` : `bottom: 100%;`}
                left: 50%;
                transform: translateX(-50%);
                transition: opacity 0.2s;
                user-select: none;
                pointer-events: none;
                text-align: center;
            `}
            className={`${theme.prefix}__slot-tooltip`}
        >
            <svg
                css={css`
                    position: absolute;
                    top: 0;
                    left: 0;
                    width: 100%;
                    height: 100%;
                    z-index: 1;
                    ${top ? `transform: rotate(180deg);` : ``}
                `}
                className={`${theme.prefix}__slot-tooltip-background`}
                viewBox="0 0 106 29"
                preserveAspectRatio="none"
            >
                <path
                    fill={theme.colors.black}
                    d="M56.75 24l-4.5 4.5-4.5-4.5H2a2 2 0 0 1-2-2V2a2 2 0 0 1 2-2h102a2 2 0 0 1 2 2v20a2 2 0 0 1-2 2H56.75z"
                />
            </svg>
            <span
                css={css`
                    position: relative;
                    top: ${top ? `16px;` : `8px;`}
                    display: block;
                    z-index: 1;
                    font-size: 12px;
                    line-height: 1;
                `}
                className={`${theme.prefix}__slot-tooltip-label`}
            >
                {time}
            </span>
        </span>
    );
};

export default Tooltip;
