// cronofy-elements/src/js/components/SlotPicker/TimeSlots.js
import React, { useContext } from "react";
import moment from "moment-timezone";
import { css, jsx } from "@emotion/core";

import Slots from "./Slots";

import { buttonReset } from "../../styles/utils";

import { I18nContext, ThemeContext, SlotsContext } from "./SlotPicker";

const TimeSlots = props => {
    const theme = useContext(ThemeContext);
    const i18n = useContext(I18nContext);
    const slots = useContext(SlotsContext);
    return (
        <div
            css={css`
                flex-grow: 1;
                padding: ${theme.padding};
                display: flex;
                flex-direction: column;
            `}
            className={`${theme.prefix}__slots-list ${theme.prefix}__slots-list--times`}
        >
            <div
                className={`${theme.prefix}__slots-header`}
                css={css`
                    display: flex;
                    justify-content: center;
                    align-items: center;
                    position: relative;
                    height: 22px;
                    margin-bottom: 10px;
                `}
            >
                <button
                    type="button"
                    css={css`
                        ${buttonReset};
                        display: block;
                        padding: 4px;
                        height: 22px;
                        width: 22px;
                        position: absolute;
                        top: 50%;
                        left: 0;
                        transform: translateY(-50%);
                    `}
                    onClick={props.backToDays}
                >
                    <svg
                        css={css`
                            fill: ${theme.colors.grey};
                            display: block;
                            width: 18px;
                            height: 18px;
                            transform: rotate(-90deg);
                        `}
                        className={`${theme.prefix}__icon ${theme.prefix}__slots-icon--back`}
                    >
                        <use xlinkHref={`#arrow`} />
                    </svg>
                    <span className="visuallyhidden">{i18n.t("back")}</span>
                </button>
                <span
                    css={css`
                        display: block;
                        padding: 0 10px 0 0;
                        line-height: 1;
                    `}
                    className={`${theme.prefix}__slots-heading`}
                >
                    {i18n.f(
                        moment(slots.day, "YYYY-MM-DDTThh:mm:00Z"),
                        `dddd Do MMMM`
                    )}
                </span>
            </div>
            <div
                css={css`
                    width: 100%;
                    flex-grow: 1;
                    display: flex;
                `}
            >
                <Slots
                    periods={slots.slots[slots.day]}
                    confirm={props.selectSlot}
                />
            </div>
        </div>
    );
};

export default TimeSlots;
