// cronofy-elements/src/js/components/generic/Error.js
import React from "react";
import { css, jsx } from "@emotion/core";
import { transparentize } from "polished";

const Error = ({ message = "There was an error", theme, title = false }) => (
    <div
        css={css`
            position: absolute;
            top: 0;
            left: 0;
            z-index: 10;
            width: 100%;
            height: 100%;
            padding: 10px;
            background: ${transparentize(0.8, theme.colors.black)};
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;
        `}
        className={`${theme.prefix}__error`}
    >
        <svg
            css={css`
                width: 2em;
                height: 2em;
                fill: ${theme.colors.black};
            `}
            className={`${theme.prefix}__error-icon`}
        >
            <use xlinkHref={`#error`} />
        </svg>
        {title ? (
            <span
                css={css`
                    font-weight: bold;
                    display: block;
                    text-align: center;
                    margin-top: 1em;
                    color: ${theme.colors.black};
                `}
            >
                {title}
            </span>
        ) : null}
        <span
            css={css`
                display: block;
                text-align: center;
                margin-top: 1em;
                max-width: 36em;
                color: ${theme.colors.black};
            `}
            className={`${theme.prefix}__error-text`}
        >
            {message}
        </span>
    </div>
);

export default Error;
