// cronofy-elements/src/js/components/generic/Loading.js
import React from "react";
import { css, jsx, keyframes } from "@emotion/core";
import { transparentize } from "polished";

import colors from "../../styles/colors";

const spin = keyframes`
        0% {
            transform: rotate(0deg);
        }
        100%{
            transform: rotate(360deg);
        }
    `;

const Loading = ({ loading, children, stylePrefix = "loading" }) =>
    loading ? (
        <div
            css={css`
                position: absolute;
                top: 0;
                left: 0;
                width: 100%;
                height: 100%;
                background: ${transparentize(0.8, colors.black)};
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: center;
            `}
            className={`${stylePrefix}__loading-wrapper`}
        >
            <svg
                css={css`
                    width: 2em;
                    height: 2em;
                    fill: ${colors.white};
                    animation: ${spin} infinite linear 1s;
                `}
                className={`${stylePrefix}__loading-icon`}
            >
                <use xlinkHref={`#reload`} />
            </svg>
        </div>
    ) : (
        children
    );

export default Loading;
