// cronofy-elements/src/js/components/generic/TransitionSlide.js
import React from "react";
import { CSSTransition, TransitionGroup } from "react-transition-group";
import { css, jsx } from "@emotion/core";

import { parseDirection } from "../../helpers/events";

const TransitionSlide = ({ direction, keyProp, children, styles }) => {
    const parsedDirection = parseDirection(direction);
    const transitionDuration = 400; // ms

    const directions = {
        forwards: { in: [`100%`, `0`], out: [`-100%`, `0`] },
        backwards: { in: [`-100%`, `0`], out: [`100%`, `0`] },
        upwards: { in: [`0`, `100%`], out: [`0`, `-100%`] },
        downwards: { in: [`0`, `-100%`], out: [`0`, `100%`] }
    };

    const in_0 = directions[parsedDirection].in[0];
    const in_1 = directions[parsedDirection].in[1];

    const out_0 = directions[parsedDirection].out[0];
    const out_1 = directions[parsedDirection].out[1];

    const XY_in = `${in_0}, ${in_1}`;
    const XY_out = `${out_0}, ${out_1}`;

    const genericTransitionStyles = css`
        position: relative;
        width: 100%;
        // height: 100%;
        display: flex;
        overflow: hidden;
        .transition--slide-enter {
            transform: translate(${XY_in});
            transition: transform ${transitionDuration}ms linear;
        }
        .transition--slide-enter-active {
            transform: translate(0, 0);
        }
        .transition--slide-exit {
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            transform: translate(0, 0);
            transition: transform ${transitionDuration}ms linear;
        }
        .transition--slide-exit-active {
            transform: translate(${XY_out});
        }
    `;
    return (
        <TransitionGroup
            css={css`
                ${genericTransitionStyles};
                ${styles};
            `}
            className={`direction--${direction}`}
        >
            <CSSTransition
                key={keyProp}
                timeout={transitionDuration}
                classNames="transition--slide"
            >
                {children}
            </CSSTransition>
        </TransitionGroup>
    );
};

export default TransitionSlide;
