// cronofy-elements/src/js/components/AvailabilityRules/AddButton.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { buttonReset } from "../../styles/utils";
import { I18nContext, ThemeContext } from "./AvailabilityRules";

const AddButton = ({ callback }) => {
    const [theme, setTheme] = useContext(ThemeContext);
    const i18n = useContext(I18nContext);
    return (
        <button
            type="button"
            css={css`
                ${buttonReset}
                background: ${theme.colors.white};
                border: 1px solid ${theme.colors.hairline};
                border-radius: 4px;
                height: 28px;
                width: 28px;
                position: relative;
                cursor: pointer;
                transition: background 0.2s;
                margin-bottom: 5px;
                &:hover {
                    background: ${theme.colors.whiteDark};
                }
            `}
            className={`${theme.prefix}__calendars__add-button`}
            onClick={() => callback()}
        >
            <svg
                css={css`
                    position: absolute;
                    top: 50%;
                    left: 50%;
                    transform: translate(-50%, -50%);
                    width: 12px;
                    height: 12px;
                    display: block;
                `}
                className={`${theme.prefix}__calendars__add-button-icon`}
                viewBox="0 0 7 6"
            >
                <path
                    css={css`
                        transform: rotate(45deg);
                        transform-origin: 50% 50%;
                    `}
                    className={`${theme.prefix}__calendars__add-button-icon--path`}
                    fill={theme.colors.black}
                    d="M3.5 2.287L5.451.336c.163-.163.455-.135.651.062.197.196.225.488.062.65L4.213 3l1.951 1.951c.163.163.135.455-.062.651-.196.197-.488.225-.65.062L3.5 3.713 1.549 5.664c-.163.163-.455.135-.651-.062-.197-.196-.225-.488-.062-.65L2.787 3 .836 1.049C.673.886.7.594.898.398c.196-.197.488-.225.65-.062L3.5 2.287z"
                />
            </svg>
            <span className="visuallyhidden">{i18n.t("add_calendar")}</span>
        </button>
    );
};

export default AddButton;
