// cronofy-elements/src/js/components/AvailabilityRules/Checkmark.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { I18nContext, ThemeContext } from "./AvailabilityRules";

const Checkmark = ({ checked }) => {
    const i18n = useContext(I18nContext);
    const [theme, setTheme] = useContext(ThemeContext);

    return (
        <div
            css={css`
                background: red;
                width: 18px;
                height: 18px;
                border-radius: 2px;
                margin-right: 8px;
                border: 1px solid
                    ${checked ? theme.colors.primary : theme.colors.hairline};
                background: ${checked
                    ? theme.colors.primary
                    : theme.colors.white};
                position: relative;
            `}
            className={`${theme.prefix}__checkbox ${theme.prefix}__checkbox--${
                checked ? i18n.t("checked") : i18n.t("unchecked")
            }`}
        >
            {checked ? (
                <svg
                    css={css`
                        position: absolute;
                        top: 50%;
                        left: 50%;
                        transform: translate(-50%, -50%);
                        width: 10px;
                        height: 10px;
                    `}
                    viewBox="0 0 7 8"
                    xmlns="http://www.w3.org/2000/svg"
                    className={`${theme.prefix}__checkbox-tickmark`}
                >
                    <path
                        d="M0 3.956l.7-.934 1.75 2.334L6.3.222l.7.934-4.55 6.066z"
                        fill={theme.colors.white}
                    />
                </svg>
            ) : null}
        </div>
    );
};

export default Checkmark;
