// cronofy-elements/src/js/components/AvailabilityRules/Loading.js
import React, { useContext } from "react";
import { css, jsx, keyframes } from "@emotion/core";
import { transparentize } from "polished";

import { I18nContext, ThemeContext } from "./AvailabilityRules";

const spin = keyframes`
    0% {
        transform: rotate(0deg);
    }
    100%{
        transform: rotate(360deg);
    }
`;

const Loading = () => {
    const [theme, setTheme] = useContext(ThemeContext);
    const i18n = useContext(I18nContext);
    return (
        <div
            css={css`
                position: absolute;
                top: 0;
                left: 0;
                width: 100%;
                height: 100%;
                background: ${transparentize(0.8, "#000000")};
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: center;
            `}
            className={`${theme.prefix}__loading-wrapper`}
        >
            <svg
                css={css`
                    width: 2em;
                    height: 2em;
                    fill: #ffffff;
                    animation: ${spin} infinite linear 1s;
                `}
            >
                <use xlinkHref={`#reload`} />
            </svg>
            <span className={"visuallyhidden"}>{i18n.t("loading")}</span>
        </div>
    );
};

export default Loading;
