// cronofy-elements/src/js/components/AvailabilityRules/RemoveButton.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { buttonReset } from "../../styles/utils";
import { I18nContext, ThemeContext } from "./AvailabilityRules";

const RemoveButton = ({ callback }) => {
    const i18n = useContext(I18nContext);
    const [theme, setTheme] = useContext(ThemeContext);
    return (
        <button
            type="button"
            css={css`
                ${buttonReset}
                margin-left: 8px;
                background: ${theme.colors.white};
                border-radius: 50%;
                height: 16px;
                width: 16px;
                position: relative;
                cursor: pointer;
                transition: background 0.2s;
                &:hover {
                    background: ${theme.colors.black};
                    .${theme.prefix}__calendars__remove-button-icon {
                        fill: ${theme.colors.white};
                    }
                }
            `}
            className={`${theme.prefix}__calendars__remove-button`}
            onClick={() => callback()}
        >
            <svg
                css={css`
                    position: absolute;
                    top: 50%;
                    left: 50%;
                    transform: translate(-50%, -50%);
                    width: 8px;
                    height: 8px;
                    display: block;
                    transition: fill 0.2s;
                    fill: ${theme.colors.black};
                `}
                className={`${theme.prefix}__calendars__remove-button-icon`}
                viewBox="0 0 7 6"
            >
                <path d="M3.5 2.287L5.451.336c.163-.163.455-.135.651.062.197.196.225.488.062.65L4.213 3l1.951 1.951c.163.163.135.455-.062.651-.196.197-.488.225-.65.062L3.5 3.713 1.549 5.664c-.163.163-.455.135-.651-.062-.197-.196-.225-.488-.062-.65L2.787 3 .836 1.049C.673.886.7.594.898.398c.196-.197.488-.225.65-.062L3.5 2.287z" />
            </svg>
            <span className="visuallyhidden">{i18n.t("remove")}</span>
        </button>
    );
};

export default RemoveButton;
