// cronofy-elements/src/js/components/AvailabilityRules/SlotsDisplay.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { SlotsContext, ThemeContext, ExtrasContext } from "./AvailabilityRules";

import SlotDisplay from "./SlotDisplay";
import { parseSlotsIntoDays } from "../../helpers/slots.rules";
import { arrayOfDaySlots } from "../../helpers/utils.AvailabilityRules";

const Slots = () => {
    const { slots, setSlots } = useContext(SlotsContext);
    const [extras, setExtras] = useContext(ExtrasContext);
    const [theme, setTheme] = useContext(ThemeContext);

    const days = parseSlotsIntoDays(slots, extras.limits.slotsPerDay);

    const daysList = days.map((day, i) => {
        const mappableSlots = arrayOfDaySlots(
            day.slots,
            i,
            extras.limits.slotsPerDay
        );
        return (
            <div
                key={day.day}
                css={css`
                    display: block;
                    margin: 0;
                    padding: 0;
                    width: ${theme.sizes.columnWidth}px;
                    flex-shrink: 1;
                    border-right: 1px solid ${theme.colors.hairline};
                    &:last-of-type {
                        border-right: none;
                    }
                `}
                className={`${theme.prefix}__slots-background-column`}
            >
                {mappableSlots.map((slot, i) => {
                    const reverseOptions = {
                        15: 4,
                        30: 2,
                        60: 1
                    };
                    const first = i < reverseOptions[extras.limits.duration];
                    return (
                        <SlotDisplay
                            key={`${day.day}_${slot.start}`}
                            slot={slot}
                            first={first}
                        />
                    );
                })}
            </div>
        );
    });

    return (
        <div
            css={css`
                display: flex;
                margin: 0;
                padding: 0;
                width: 100%;
                align-content: stretch;
                position: absolute;
                top: 0;
                left: 0;
            `}
            className={`${theme.prefix}__slots-background-columns`}
        >
            {daysList}
        </div>
    );
};

export default Slots;
