// cronofy-elements/src/js/components/CalendarSync/Active.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { I18nContext, ThemeContext } from "./CalendarSync";

const Active = () => {
    const i18n = useContext(I18nContext);
    const theme = useContext(ThemeContext);

    return (
        <div
            css={css`
                display: flex;
                align-items: center;
                padding: 10px 0;
            `}
            className={`${theme.prefix}__status--active`}
        >
            <span
                css={css`
                    display: block;
                    color: ${theme.colors.greyMid};
                    text-transform: uppercase;
                    font-weight: bold;
                    letter-spacing: 1px;
                    font-size: 10px;
                    line-height: 1;
                `}
                className={`${theme.prefix}__status-label`}
            >
                {i18n.t("active")}
            </span>
            <svg
                css={css`
                    display: block;
                    width: 16px;
                    height: 16px;
                    margin-right: 4px;
                    margin-left: 4px;
                    background: ${theme.colors.green};
                    border-radius: 50%;
                `}
                viewBox="0 0 14 15"
                className={`${theme.prefix}__status-icon ${theme.prefix}__status-icon--tick`}
            >
                <path
                    css={css`
                        fill: ${theme.colors.white};
                    `}
                    d="M3.095 7.892l.65-.927L6.06 8.587l3.57-5.097.926.649-4.218 6.024z"
                />
            </svg>
        </div>
    );
};

export default Active;
