// cronofy-elements/src/js/components/CalendarSync/AddToggle.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { buttonReset } from "../../styles/utils";
import { I18nContext, ThemeContext, StatusContext } from "./CalendarSync";

const AddToggle = () => {
    const theme = useContext(ThemeContext);
    const [status, setStatus] = useContext(StatusContext);
    const i18n = useContext(I18nContext);

    const handleAddToggle = () => {
        setStatus({ ...status, adding: !status.adding });
    };

    return (
        <div
            css={css`
                height: 40px;
                display: flex;
                align-items: center;
                margin-right: 10px;
            `}
            className={`${theme.prefix}__add-toggle`}
        >
            <button
                type="button"
                css={css`
                    ${buttonReset}
                    padding: 0 10px;
                    height: 30px;
                    display: flex;
                    align-items: center;
                    cursor: pointer;
                    border: 1px solid ${theme.colors.hairline};
                    border-radius: 4px;
                `}
                className={`${theme.prefix}__add-toggle-button`}
                onClick={handleAddToggle}
            >
                <svg
                    css={css`
                        display: block;
                        width: 12px;
                        height: 12px;
                        margin-right: 10px;
                        fill: ${theme.colors.grey};
                    `}
                    className={`${theme.prefix}__add-toggle-icon`}
                    viewBox={status.adding ? "0 0 13 10" : "0 0 13 13"}
                >
                    {status.adding ? (
                        <path d="M2.756 4.41L6.333.834 5.5 0l-5 5 .417.417L5.5 10l.833-.833-3.577-3.578h9.53V4.411z" />
                    ) : (
                        <path d="M7.319 5.417h4.482c.374 0 .677.367.677.82 0 .452-.303.818-.677.818H7.32v4.483c0 .374-.367.677-.819.677-.452 0-.819-.303-.819-.677V7.055H1.199c-.374 0-.677-.366-.677-.819 0-.452.303-.819.677-.819H5.68V.935c0-.374.367-.677.819-.677.452 0 .819.303.819.677v4.482z" />
                    )}
                </svg>
                <span
                    css={css`
                        line-height: 1;
                        font-weight: bold;
                    `}
                    className={`${theme.prefix}__add-toggle-text`}
                >
                    {status.adding
                        ? i18n.t("back")
                        : i18n.t("add_calendar_account")}
                </span>
            </button>
        </div>
    );
};

export default AddToggle;
