// cronofy-elements/src/js/components/CalendarSync/EditToggle.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { buttonReset } from "../../styles/utils";
import { I18nContext, ThemeContext, StatusContext } from "./CalendarSync";

const EditToggle = () => {
    const [status, setStatus] = useContext(StatusContext);
    const theme = useContext(ThemeContext);
    const i18n = useContext(I18nContext);

    const handleEditToggle = () => {
        setStatus({ ...status, editing: !status.editing });
    };

    const buttonTextStyles = status.editing
        ? css`
              border-bottom: 2px solid ${theme.colors.primary};
          `
        : css`
              color: ${theme.colors.greyMid};
          `;

    return (
        <div
            css={css`
                height: 40px;
                display: flex;
                align-items: center;
            `}
            className={`${theme.prefix}__edit-toggle`}
        >
            <button
                type="button"
                css={css`
                    ${buttonReset}
                    padding: 0;
                    cursor: pointer;
                `}
                className={`${theme.prefix}__edit-toggle-button`}
                onClick={handleEditToggle}
            >
                {status.editing ? (
                    <svg
                        css={css`
                            width: 10px;
                            height: 10px;
                            margin-right: 4px;
                            fill: ${theme.colors.grey};
                        `}
                        className={`${theme.prefix}__edit-toggle-icon`}
                        viewBox="0 0 13 10"
                    >
                        <path d="M2.756 4.41L6.333.834 5.5 0l-5 5 .417.417L5.5 10l.833-.833-3.577-3.578h9.53V4.411z" />
                    </svg>
                ) : null}
                <span
                    css={css`
                        font-weight: bold;
                        ${buttonTextStyles}
                    `}
                    className={`${theme.prefix}__edit-toggle-text`}
                >
                    {status.editing
                        ? i18n.t("cancel")
                        : i18n.t("edit_accounts")}
                </span>
            </button>
        </div>
    );
};

export default EditToggle;
