// cronofy-elements/src/js/components/CalendarSync/Inactive.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import { I18nContext, ThemeContext } from "./CalendarSync";

const Inactive = ({ url }) => {
    const theme = useContext(ThemeContext);
    const i18n = useContext(I18nContext);

    return (
        <a
            href={url}
            target="_blank"
            css={css`
                display: flex;
                align-items: center;
                padding: 2px 6px;
                background: ${theme.colors.red};
                border-radius: 4px;
                transition: background 0.2s;
                &:hover {
                    background: ${theme.colors.redDark};
                }
                @media (max-width: 380px) {
                    margin-top: 6px;
                    padding: 6px 10px;
                    width: 100px;
                    justify-content: center;
                }
            `}
            className={`${theme.prefix}__status--inactive`}
        >
            <span
                css={css`
                    display: block;
                    color: ${theme.colors.white};
                    text-transform: uppercase;
                    font-weight: bold;
                    letter-spacing: 1px;
                    font-size: 10px;
                    vertical-align: middle;
                    padding: 0 4px;
                `}
                className={`${theme.prefix}__status-label`}
            >
                {i18n.t("relink")}
            </span>
            <svg
                css={css`
                    display: block;
                    width: 13px;
                    height: 13px;
                    margin-left: 4px;
                    fill: ${theme.colors.white};
                `}
                className={`${theme.prefix}__status-icon ${theme.prefix}__status-icon--reload`}
                viewBox="0 0 9 13"
                xmlns="http://www.w3.org/2000/svg"
            >
                <path d="M4.459 2.3v1.2a3 3 0 0 0-2.124 5.124l-.846.846a4.2 4.2 0 0 1 2.97-7.17zm2.97 1.23a4.2 4.2 0 0 1-2.97 7.17V9.5a3 3 0 0 0 2.124-5.124l.846-.846zm-2.97 8.97l-2.4-2.4 2.4-2.4v4.8zm0-7.2V.5l2.4 2.4-2.4 2.4z" />
            </svg>
        </a>
    );
};

export default Inactive;
