// cronofy-elements/src/js/components/CalendarSync/Pending.js
import React, { useContext } from "react";
import { css, jsx, keyframes } from "@emotion/core";

import { I18nContext, ThemeContext } from "./CalendarSync";

const Pending = () => {
    const theme = useContext(ThemeContext);
    const i18n = useContext(I18nContext);

    const spin = keyframes`
        0% {
            transform: rotate(0deg);
        }
        100%{
            transform: rotate(360deg);
        }
    `;

    return (
        <span
            css={css`
                display: flex;
                align-items: center;
                padding: 4px;
                border-radius: 4px;
            `}
            className={`${theme.prefix}__status--pending`}
        >
            <span
                css={css`
                    display: block;
                    color: ${theme.colors.greyMid};
                    text-transform: uppercase;
                    font-weight: bold;
                    letter-spacing: 1px;
                    font-size: 10px;
                    vertical-align: middle;
                `}
                className={`${theme.prefix}__status-label`}
            >
                {i18n.t("pending")}
            </span>
            <svg
                css={css`
                    display: block;
                    width: 16px;
                    height: 16px;
                    margin-left: 4px;
                    fill: ${theme.colors.white};
                    background: ${theme.colors.yellow};
                    padding: 2px;
                    border-radius: 50%;
                `}
                className={`${theme.prefix}__status-icon ${theme.prefix}__status-icon--pending`}
                viewBox="0 0 98 98"
            >
                <path d="M49 97.074C22.45 97.074.926 75.55.926 49 .926 22.45 22.45.926 49 .926 75.55.926 97.074 22.45 97.074 49c0 26.55-21.524 48.074-48.074 48.074zm0-9.615c21.24 0 38.459-17.219 38.459-38.459 0-21.24-17.219-38.459-38.459-38.459-21.24 0-38.459 17.219-38.459 38.459 0 21.24 17.219 38.459 38.459 38.459zM44 20h10v31H44V20z" />
                <rect
                    x="44"
                    y="49"
                    width="10"
                    height="31"
                    css={css`
                        transform: rotate(-45deg);
                        transform-origin: 50% 50%;
                        animation: ${spin} infinite linear 2s;
                    `}
                />
            </svg>
        </span>
    );
};

export default Pending;
