// cronofy-elements/src/js/components/CalendarSync/Providers.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";

import {
    LoggingContext,
    ThemeContext,
    ProvidersContext,
    StatusContext
} from "./CalendarSync";
import ProviderLogo from "./ProviderLogo";

const Provider = ({ provider }) => {
    const log = useContext(LoggingContext);
    const theme = useContext(ThemeContext);
    const [status, setStatus] = useContext(StatusContext);

    const handleProviderLink = e => {
        if (status.demo) {
            e.preventDefault();
            log.info(
                "Demo mode is active, so the authorization links have been disabled."
            );
        }
    };

    return (
        <a
            css={css`
                border: 1px solid ${theme.colors.hairline};
                padding: 4px 10px;
                border-radius: 4px;
                display: flex;
                align-items: center;
                margin-right: 10px;
                margin-bottom: 10px;
                transition: background 0.2s;
                cursor: pointer;
                color: inherit;
                width: 164px;
                background: ${theme.colors.white};
                &:hover {
                    background: ${theme.colors.greyPale};
                }
            `}
            className={`${theme.prefix}__provider-auth`}
            href={provider.authlink}
            onClick={e => handleProviderLink(e)}
        >
            <ProviderLogo provider={provider.slug} fill={true} size={36} />
            <span
                className={`${theme.prefix}__provider-auth-text`}
                css={css`
                    display: block;
                    flex-grow: 1;
                    text-align: center;
                `}
            >
                {provider.name}
            </span>
        </a>
    );
};

const Providers = () => {
    const theme = useContext(ThemeContext);
    const providers = useContext(ProvidersContext);

    const providersList = providers.map(provider => (
        <Provider key={provider.slug} provider={provider} />
    ));

    return (
        <div
            css={css`
                display: flex;
                flex-wrap: wrap;
                align-items: space-between;
            `}
            className={`${theme.prefix}__providers`}
        >
            {providersList}
        </div>
    );
};

export default Providers;
