// cronofy-elements/stories/story-temps/mock-data.js
import moment from "moment";

export const today = moment(new Date());

export const AVAILABLE_PERIODS = [
    {
        available: true,
        start: `${today.format(`YYYY-MM-DD`)}T09:00:00Z`,
        end: `${today.format(`YYYY-MM-DD`)}T10:00:00Z`
    },
    {
        available: true,
        start: `${today.format(`YYYY-MM-DD`)}T10:00:00Z`,
        end: `${today.format(`YYYY-MM-DD`)}T11:00:00Z`
    },
    {
        available: true,
        start: `${today.format(`YYYY-MM-DD`)}T11:00:00Z`,
        end: `${today.format(`YYYY-MM-DD`)}T12:00:00Z`
    },
    {
        available: true,
        start: `${today.format(`YYYY-MM-DD`)}T13:00:00Z`,
        end: `${today.format(`YYYY-MM-DD`)}T14:00:00Z`
    },
    {
        available: true,
        start: `${today.format(`YYYY-MM-DD`)}T14:00:00Z`,
        end: `${today.format(`YYYY-MM-DD`)}T15:00:00Z`
    }
];

const days = today;
export const AVAILABLE_DAYS = [
    days.format(`YYYY-MM-DD`),
    days.add(1, "days").format(`YYYY-MM-DD`),
    days.add(1, "days").format(`YYYY-MM-DD`),
    // days.add(1, "days").format(`YYYY-MM-DD`),
    // days.add(1, "days").format(`YYYY-MM-DD`),
    // days.add(1, "days").format(`YYYY-MM-DD`),
    days.add(1, "days").format(`YYYY-MM-DD`)
];
