// cronofy-elements/stories/story-temps/TransitionToggler.js
import React from "react";
import { css } from "@emotion/core";

import TransitionSlide from "../../src/js/components/generic/TransitionSlide";

class TransitionToggler extends React.Component {
    state = {
        view: 1
    };
    toggleView(target) {
        this.setState({ view: target });
    }
    render() {
        return (
            <TransitionSlide
                direction={this.props.direction}
                keyProp={this.state.view}
            >
                {this.state.view === 1 ? (
                    <div
                        css={css`
                            background: red;
                            width: 100%;
                            height: 100%;
                        `}
                    >
                        <button onClick={() => this.toggleView(2)}>
                            show 2
                        </button>
                    </div>
                ) : (
                    <div
                        css={css`
                            background: blue;
                            width: 100%;
                            height: 100%;
                        `}
                    >
                        <button onClick={() => this.toggleView(1)}>
                            show 1
                        </button>
                    </div>
                )}
            </TransitionSlide>
        );
    }
}

export default TransitionToggler;
