// cronofy-elements/stories/story-temps/Wrapper.js
import React from "react";
import { css } from "@emotion/core";

import globals from "../../src/js/styles/global";
import colors from "../../src/js/styles/colors";

const Wrapper = ({ layout = [326, 400], children }) => {
    const containerStyles = css`
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
        background-color: rgba(0, 0, 0, 0.05);
        background-image: repeating-linear-gradient(
                0deg,
                transparent,
                transparent 7px,
                rgba(0, 0, 0, 0.2) 1px,
                transparent 8px
            ),
            repeating-linear-gradient(
                90deg,
                transparent,
                transparent 7px,
                rgba(0, 0, 0, 0.2) 1px,
                transparent 8px
            );
        background-size: 8px 8px;
    `;
    const wrapperStyles = css`
        ${globals};
        width: ${layout[0]}px;
        height: ${layout[1]}px;
        // border: 1px solid red;
        background: ${colors.white};
        // padding: 1.5em;
        font-family: sans-serif;
        // display: flex;
        margin: 1em;
        // flex-direction: column;
    `;
    return (
        <div css={containerStyles}>
            <div css={wrapperStyles}>{children}</div>
        </div>
    );
};

export default Wrapper;
