// cronofy-elements/src/js/components/Agenda/EventWrapper.js
import React, { useContext } from "react";
import { css, jsx } from "@emotion/core";
import { transparentize } from "polished";

import { getCalendarColors } from "../../helpers/colors";

import { ThemeContext } from "./Agenda";

const EventWrapper = props => {
    const theme = useContext(ThemeContext);
    let colors = getCalendarColors(theme.colors, props.calIndex);

    const prefix = theme.prefix;

    const transparency = props.transparency || props.data.transparency;
    return (
        <div
            css={css`
                background-color: ${colors.main};
                color: ${colors.text};
                padding: 4px 10px 14px 10px;
                margin-bottom: 1em;
                position: relative;
                cursor: pointer;
                border-left: 6px solid ${colors.dark};
                transition: background-color 0.2s, box-shadow 0.2s, color 0.2s;
                
                &:last-of-type {
                    margin-bottom: 0;
                }
                
                &:not(.${prefix}__event--expanded):hover {
                    background-color: ${colors.mid};
                }
        
                &.${prefix}__event--line {
                    padding: 0.3em 1em;
                    margin-bottom: 10px;
                }
        
                &.${prefix}__event--needs_action {
                    background: ${theme.colors.white};
                    color: ${theme.colors.black};
                    padding-top: 10px;
                    border-top: 1px solid ${colors.dark}; 
                    border-right: 1px solid ${colors.dark};
                    border-bottom: 1px solid ${colors.dark};
                    background-image: repeating-linear-gradient(
                        -45deg,
                        transparent,
                        transparent 8px,
                        ${transparentize(0.9, colors.dark)} 8px,
                        ${transparentize(0.9, colors.dark)} 16px
                    );
                }

                &.${prefix}__event--needs_action:not(.${prefix}__event--expanded):hover {
                    color: ${colors.text};
                }

                &.${prefix}__event--needs_action:not(.${prefix}__event--expanded):hover
                    .${prefix}__event__message--participation {
                    background: ${colors.mid};
                    color: ${transparentize(0.6, colors.text)};
                }
        
                .${prefix}__event__message {
                    position: absolute;
                    top: 0;
                    left: 50%;
                    transform: translate(-50%, 0);
                    user-select: none;
                }
        
                &.${prefix}__event--needs_action .${prefix}__event__message--participation {
                    display: block;
                    text-transform: uppercase;
                    letter-spacing: 0.1em;
                    // color: ${transparentize(0.6, colors.text)};
                    // background: ${colors.main};
                    color: ${transparentize(0.6, colors.black)};
                    background: ${colors.white};
                    padding: 0.2em 0.8em 0.2em 0.8em;
                    font-size: 0.6em;
                    transition: background 0.2s;
                }
        
                .${prefix}__event__times {
                    font-size: 0.8em;
                }
        
                &.${prefix}__event--transparent {
                    border-left: 0;
                }
        
                .${prefix}__event__message--transparency {
                    opacity: 0;
                    display: block;
                    top: auto;
                    bottom: 2px;
                    left: 50%;
                    text-transform: uppercase;
                    letter-spacing: 0.1em;
                    font-size: 0.6em;
                    transition: opacity 0.2s;
                }
                &:hover .${prefix}__event__message--transparency {
                    opacity: 1;
                }
        
                .${prefix}__event__times--time {
                    text-transform: lowercase;
                }
                .${prefix}__event__title {
                    display: block;
                    font-weight: bold;
                    margin-top: 0.2em;
                }
                &.${prefix}__event--line .${prefix}__event__title {
                    display: inline;
                    margin-top: 0;
                    margin-right: 0.5em;
                }
                .${prefix}__event__description {
                    font-size: 0.8em;
                    white-space: pre-line;
                }
        
                .${prefix}__event__meta {
                    position: absolute;
                    top: 0.5em;
                    right: 0.5em;
                    display: flex;
                    align-items: center;
                    justify-content: flex-end;
                }
        
                .${prefix}__event__attendees {
                    margin-top: 0.2em;
                    margin-left: 0.5em;
                }
        
                .${prefix}__before-icon,
                .${prefix}__after-icon {
                    position: absolute;
                    top: 0;
                    left: 100%;
                    height: 100%;
                    width: 0.5em;
                    fill: ${colors.main};
                    transition: fill 0.2s;
                }
                .${prefix}__before-icon {
                    left: 0;
                    transform: translate(-0.8em, 0) rotate(180deg);
                }
                &:hover .${prefix}__after-icon,
                &:hover .${prefix}__before-icon {
                    fill: ${colors.mid};
                }
        
                &.${prefix}__event--transparent .${prefix}__before-icon {
                    transform: translate(-0.5em, 0) rotate(180deg);
                }
                &.${prefix}__event--has-before {
                    margin-left: 0.5em;
                }
                &.${prefix}__event--has-after {
                    margin-right: 0.5em;
                }
        
                // "Recurring Event" icon
                .${prefix}__event__icon--recurring {
                    display: inline-block;
                    margin-left: 0.5em;
                    width: 10px;
                    height: 10px;
                    border: 1px solid ${colors.text};
                    border-radius: 0.3em;
                    position: relative;
                    &:before {
                        content: "";
                        display: block;
                        width: 6px;
                        height: 6px;
                        position: absolute;
                        top: 0;
                        left: 50%;
                        transform: translate(-50%, -50%) rotate(45deg);
                        border-top: 2px solid ${colors.main};
                        border-right: 2px solid ${colors.main};
                        box-shadow: inset ${colors.text} -1px 1px 0 0;
                        transition: border-color 0.2s;
                    }
                }
                &:not(.${prefix}__event--expanded):hover .${prefix}__event__icon--recurring:before {
                    border-top-color: ${colors.mid};
                    border-right-color: ${colors.mid};
                }
        
                // "Attendees" icon
                .${prefix}__event__attendees {
                    // position: absolute;
                    // bottom: 0.2em;
                    // right: 0.5em;
                }
                .${prefix}__event__attendee {
                    display: inline-block;
                    width: 12px;
                    height: 8px;
                    background: ${colors.text};
                    margin-left: -5px;
                    border-top-left-radius: 10px 10px;
                    border-top-right-radius: 10px 10px;
                    border: 1px solid ${colors.main};
                    position: relative;
                    &:before {
                        content: "";
                        display: block;
                        position: absolute;
                        top: 0;
                        left: 50%;
                        height: 8px;
                        width: 8px;
                        background: ${colors.text};
                        border-radius: 50%;
                        transform: translate(-50%, -7px);
                        border: 1px solid ${colors.main};
                    }
                }
                &:not(.${prefix}__event--expanded):hover .${prefix}__event__attendee {
                    border-color: ${colors.mid};
                    &:before {
                        border-color: ${colors.mid};
                    }
                }
                .${prefix}__event__attendee-count {
                    font-size: 0.8em;
                    position: absolute;
                    bottom: 100%;
                    right: 0;
                    opacity: 0;
                    background: ${colors.black};
                    color: ${colors.white};
                    width: 1.2em;
                    height: 1.2em;
                    display: block;
                    padding: 0.1em 0 0 0;
                    line-height: 1;
                    border-radius: 0.6em;
                    text-align: center;
                }
                .${prefix}__event__attendees:hover .${prefix}__event__attendee-count {
                    opacity: 1;
                }`}
            className={`
                ${prefix}__event
                ${prefix}__event--${transparency}
                ${prefix}__event--${props.data.participation_status}
                ${typeof props.className !== "undefined" ? props.className : ""}
            `}
            onClick={() => props.expand(props.data.event_uid)}
            title={props.data.summary}
        >
            {props.children}
        </div>
    );
};

export default EventWrapper;
